text/3486-feature-deprecation.md

```diff
@@ -0,0 +1,159 @@
+
+Cargo features are widely used and typically have lifecycles the same as other
+API components. There is not currently a way to indicate that a feature is
+intended for removal and warn about it: This RFC proposes a `deprecated` key
+that shows this information.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A new `deprecated` key will be allowed for features, defaulting to `false` if
+not specified. If specified, the value can be either a boolean, a string, or an
+object with `since` and/or `note` keys. Cargo will warn downstream crates using
+this feature.
+
+```toml
+[features]
+foo = { enables = [], deprecated = true }
+foo = { enables = [], deprecated = "this works as a note" }
+bar = { enables = [], deprecated = { since = "1.2.3", note = "don't use this!" } }
+```
+
+See [`feature-metadata`] for information about `enables`.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+- If a boolean value, `false` indicates not deprecated and `true` indicates
+  deprecated
+- If an object, the keys `since` and/or `note` can be specified
+  - An empty object is not allowed to avoid ambiguity `foo = { deprecated = {} }`
+- If a string (e.g. `foo = { deprecated = "my msg" }`), it will be equivalent to if that
+  string was specified in the `note` field (e.g. `foo = { deprecated = { note = "my msg" } }`)
+- If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning that contains the `note`. This warning should not be
+emitted for crates that reexport the feature under a feature also marked
+deprecated. For example: crate `foo` exports feature `phooey`, and crate `bar`
+exports feature `barred = ["foo/phooey"]`. If `foo` markes `phooey` as deprecated,
+running any cargo action on `bar` will emit a warning unless `barred` is also
+marked `deprecated`.
+
+Accessing this information will require access to the manifest as it will not be
+in the index.
+
+## A note on `since`
+
+The exact behavior of the `since` key is not provided in this RFC as there are
+decisions related to resolution that need to be made. The generally accepted
+concept is that there should be a warning if a deprecated feature is used _and_
+there is something actionable to resolve this issue for all downstream crates -
+but the details of how best to do this are not yet clear. Please see
+[discussion on since].
+
+If the exact behavior of `since` does not reach consensus before `deprecated` is
+nearing stabilization, this key can stabilized separately or dropped entirely.
+
+## Index changes
+
+[index changes]: #index-changes
+
+The infromation provided by `deprecated` needs to be stored in the index, and
```

Comment:
This section contradicts the sentence above that says the information will not be in the index. It's not clear why this is suggesting it should be in the index. Offhand I don't know why this would be added, since I don't think the resolver needs to know this.